drivers/dac/ad3552r/ad3552r.c

```diff
@@ -0,0 +1,1370 @@
+	},
+	[AD3552R_VREF_SELECT] = {
+		AD3552R_REG_ADDR_SH_REFERENCE_CONFIG,
+		AD3552R_MASK_REFERENCE_VOLTAGE_SEL
+	},
+	[AD3552R_CRC_ENABLE] = {
+		AD3552R_REG_ADDR_INTERFACE_CONFIG_C,
+		AD3552R_MASK_CRC_ENABLE
+	},
+#if AD3552R_QSPI_IMPLEMENTED
+	[AD3552R_SPI_MULTI_IO_MODE] = {
+		AD3552R_REG_ADDR_TRANSFER_REGISTER,
+		AD3552R_MASK_MULTI_IO_MODE
+	},
+	[AD3552R_SPI_DATA_RATE] = {
+		AD3552R_REG_ADDR_INTERFACE_CONFIG_D,
+		AD3552R_MASK_SPI_CONFIG_DDR
+	},
+	[AD3552R_SPI_SYNCHRONOUS_ENABLE] = {
+		AD3552R_REG_ADDR_INTERFACE_CONFIG_D,
+		AD3552R_MASK_DUAL_SPI_SYNCHROUNOUS_EN
+	},
+#endif
+};
+
+/* 0 -> reg addr, 1->ch0 mask, 2->ch1 mask */
+static const uint16_t addr_mask_map_ch[][3] = {
+	[AD3552R_CH_DAC_POWERDOWN] = {
+		AD3552R_REG_ADDR_POWERDOWN_CONFIG,
+		AD3552R_MASK_CH_DAC_POWERDOWN(0),
+		AD3552R_MASK_CH_DAC_POWERDOWN(1)
+	},
+	[AD3552R_CH_AMPLIFIER_POWERDOWN] = {
+		AD3552R_REG_ADDR_POWERDOWN_CONFIG,
+		AD3552R_MASK_CH_AMPLIFIER_POWERDOWN(0),
+		AD3552R_MASK_CH_AMPLIFIER_POWERDOWN(1)
+	},
+	[AD3552R_CH_OUTPUT_RANGE_SEL] = {
+		AD3552R_REG_ADDR_CH0_CH1_OUTPUT_RANGE,
+		AD3552R_MASK_CH_OUTPUT_RANGE_SEL(0),
+		AD3552R_MASK_CH_OUTPUT_RANGE_SEL(1)
+	},
+	[AD3552R_CH_TRIGGER_SOFTWARE_LDAC] = {
+		AD3552R_REG_ADDR_SW_LDAC_16B,
+		AD3552R_MASK_CH(0),
+		AD3552R_MASK_CH(1)
+	},
+	[AD3552R_CH_HW_LDAC_MASK] = {
+		AD3552R_REG_ADDR_HW_LDAC_16B,
+		AD3552R_MASK_CH(0),
+		AD3552R_MASK_CH(1)
+	},
+	[AD3552R_CH_SELECT] = {
+		AD3552R_REG_ADDR_CH_SELECT_16B,
+		AD3552R_MASK_CH(0),
+		AD3552R_MASK_CH(1)
+	},
+};
+
+static const int32_t ad3542r_ch_ranges[][2] = {
+	[AD3542R_CH_OUTPUT_RANGE_0__2P5V]	= {0, 2500},
+	[AD3542R_CH_OUTPUT_RANGE_0__3V]		= {0, 3000},
+	[AD3542R_CH_OUTPUT_RANGE_0__5V]		= {0, 5000},
+	[AD3542R_CH_OUTPUT_RANGE_0__10V]	= {0, 10000},
+	[AD3542R_CH_OUTPUT_RANGE_NEG_2P5__7P5V]	= {-2500, 7500},
+	[AD3542R_CH_OUTPUT_RANGE_NEG_5__5V]	= {-5000, 5000}
+};
+
+static const int32_t ad3552r_ch_ranges[][2] = {
+	[AD3552R_CH_OUTPUT_RANGE_0__2P5V]	= {0, 2500},
+	[AD3552R_CH_OUTPUT_RANGE_0__5V]		= {0, 5000},
+	[AD3552R_CH_OUTPUT_RANGE_0__10V]	= {0, 10000},
+	[AD3552R_CH_OUTPUT_RANGE_NEG_5__5V]	= {-5000, 5000},
+	[AD3552R_CH_OUTPUT_RANGE_NEG_10__10V]	= {-10000, 10000}
+};
+
+static const int32_t gains_scaling_table[] = {
+	[AD3552R_CH_GAIN_SCALING_1]		= 1000,
+	[AD3552R_CH_GAIN_SCALING_0_5]		= 500,
+	[AD3552R_CH_GAIN_SCALING_0_25]		= 250,
+	[AD3552R_CH_GAIN_SCALING_0_125]		= 125
+};
+
+static const uint16_t ad3552r_chip_ids[] = {
+	[AD3542R_ID] = 0x4008,
+	[AD3552R_ID] = 0x4009,
+};
+
+static uint8_t _ad3552r_reg_len(uint8_t addr)
+{
+	switch (addr) {
+	case AD3552R_REG_ADDR_HW_LDAC_16B:
+	case AD3552R_REG_ADDR_CH_SELECT_16B:
+	case AD3552R_REG_ADDR_SW_LDAC_16B:
+	case AD3552R_REG_ADDR_HW_LDAC_24B:
+	case AD3552R_REG_ADDR_CH_SELECT_24B:
+	case AD3552R_REG_ADDR_SW_LDAC_24B:
+		return 1;
+	default:
+		break;
+	}
+
+	if (addr > AD3552R_REG_ADDR_HW_LDAC_24B)
+		return 3;
+	if (addr > AD3552R_REG_ADDR_HW_LDAC_16B)
+		return 2;
+
+	return 1;
```

Comment:
nitpick: would only this part of the function be enough?

Reply:
No, the registers in the switch are special cases that don't fit in the if